src/main/kotlin/app/revanced/patcher/fingerprint/method/impl/MethodFingerprint.kt

```diff
@@ -42,6 +44,93 @@ abstract class MethodFingerprint(
     var result: MethodFingerprintResult? = null
 
     companion object {
+        private fun MethodFingerprint.signatureKey(): String {
+            if (accessFlags == null) return ""
+
+            var returnTypeValue = returnType
+            if (returnTypeValue == null) {
+                if (accessFlags and AccessFlags.CONSTRUCTOR.value != 0) {
+                    // Constructors always have void return type
+                    returnTypeValue = "V"
+                } else {
+                    return ""
+                }
+            }
+
+            return buildString {
+                append(accessFlags)
+                append(returnTypeValue.first())
+                parameters?.forEach { append(it.first()) }
+            }
+        }
+
+        internal fun createMethodLookupMap(classes: Iterable<ClassDef>) : Map<String, List<Pair<ClassDef, Method>>> {
+            fun addMethodToMapList(map: MutableMap<String, MutableList<Pair<ClassDef, Method>>>,
+                                   key:String, classDef : ClassDef, method: Method) {
+                var list = map[key]
+                if (list == null) {
+                    list = LinkedList()
+                    map[key] = list
+                }
+                list += Pair(classDef, method)
+            }
+
+            val methodMap = mutableMapOf<String, MutableList<Pair<ClassDef, Method>>>()
+            for (classDef in classes) {
+                for (method in classDef.methods) {
+                    val accessFlagsReturnKey = method.accessFlags.toString() + method.returnType.first()
+                    val accessFlagsReturnParametersKey = buildString {
+                        append(accessFlagsReturnKey)
+                        method.parameterTypes.forEach { append(it.first()) }
+                    }
+
+                    // For signatures with no access or return type specified
+                    addMethodToMapList(methodMap, "", classDef, method)
+                    // Only access and return type
+                    addMethodToMapList(methodMap, accessFlagsReturnKey, classDef, method)
+                    // Access, return, and parameter count
+                    addMethodToMapList(methodMap, accessFlagsReturnParametersKey, classDef, method)
+                    // Using more fine grained lookup is possible, but in practice it does not improve performance
+
```

Comment:
I don't know why the map is designed this way. A single map is used for everything and is missing some proper structure/abstraction. Instead, a lookup table can be created for each type of key.

Reply:
That sounds good. I am not yet entirely sure on how certain abstractions should look like for this PR, but this should work to proceed for now.